File: programmers/PART14/기능개발.py
import math


def solution(progresses, speeds):
    '''
    [93, 30, 55] [1, 30, 5]

                    | Ceiling
    100 - 93 = 7    | 7 / 1 = 7           : 7일 후 완료
    100 - 30 = 70   | 70 / 30 = 2.33..    : 3일 후 완료 (7일 후 동시 배포)
    100 - 55 = 45   | 45 / 5 = 9          : 9일 후 완료

    7일 후: 2개 기능 배포
    9일 후: 1개 기능 배포


    [95, 90, 99, 99, 80, 99], [1, 1, 1, 1, 1, 1]

                    | Ceiling
    100 - 95 = 5    | 5 / 1 = 5            : 5일 후 완료
    100 - 90 = 10   | 10 / 1 = 10          : 10일 후 완료
    100 - 99 = 1    | 1 / 1 = 1            : 1일 후 완료 (10일 후 동시 배포)
    100 - 99 = 1    | 1 / 1 = 1            : 1일 후 완료 (10일 후 동시 배포)
    100 - 80 = 20   | 20 / 1 = 20          : 20일 후 완료
    100 - 90 = 10   | 10 / 1 = 10          : 10일 후 완료 (20일 후 동시 배포)

    5일 후: 1개 기능 배포
    10일 후: 3개 기능 배포
    20일 후: 2개 기능 배포

    answer = [1]

    available [5, 10, 1, 1, 20, 10]
    last_day = 5

    last_day < 10 : answer = [1, 1]
    last_day = 10

    last_day > 1 : answer = [1, 2]
    last_day > 1 : answer = [1, 3]

    last_day < 20 : answer = [1, 3, 1]
    last_day = 20

    last_day > 10 : answer = [1, 3, 2]
    '''
    available = []
    answer = [1]
    last_index = 0

    for p, s in zip(progresses, speeds):
        available.append(math.ceil((100 - p) / s))

    last_day = available[0]
    for a in available[1:]:
        if last_day < a:
            answer.append(1)
            last_index += 1
            last_day = a
        else:
            answer[last_index] += 1

    return answer


# print(solution([93, 30, 55], [1, 30, 5]))
print(solution([95, 90, 99, 99, 80, 99], [1, 1, 1, 1, 1, 1]))

File: programmers/PART14/기능개발.cpp
#include <iostream>

#include <string>
#include <vector>
#include <cmath>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds)
{
  vector<int> answer{ 1 };
  vector<int> available;
  int last_index = 0;
  const int size = progresses.size();

  for (int i = 0; i < size; ++i) {
    const int progress = progresses.at(i);
    const int speed = speeds.at(i);
    available.push_back(ceil((100 - progress) / (float)speed));
  }

  int last_day = available.at(0);
  for (int i = 1; i < size; ++i) {
    const int day = available.at(i);
    if (last_day < day) {
      answer.push_back(1);
      last_index += 1;
      last_day = day;
    } else {
      answer[last_index] += 1;
    }
  }

  return answer;
}

int main()
{
  vector<int> progresses{ 95, 90, 99, 99, 80, 90 };
  vector<int> speeds{ 1, 1, 1, 1, 1, 1 };
  vector<int> result = solution(progresses, speeds);

  for (auto i : result) {
    cout << i << ' ';
  }

  return 0;
}